homeassistant/components/hive/binary_sensor.py

```diff
@@ -113,12 +113,15 @@ async def async_update(self) -> None:
         await self.hive.session.updateData(self.device)
         self.device = await self.hive.sensor.getSensor(self.device)
         self.attributes = self.device.get("attributes", {})
-        self._attr_is_on = self.device["status"]["state"]
+
         if self.device["hiveType"] != "Connectivity":
             self._attr_available = self.device["deviceData"].get("online")
         else:
             self._attr_available = True
 
+        if self._attr_available:
```

Comment:
I'm confused, `async_update` isn't called when the sensor is unavailable. So, the problem seems to be somewhere else, as this case you trying to fix, should not happen in the first place?

Are you sure available status is set correctly? Should this check for the state begin present in the device status be part of the determination of a device being available or not?

Reply:
This happens after a restart of  home assistant so at start up or on first setup of the integration.

The scenario is I have multiple door sensors in my house more than the amount of doors I have so the one that has ran out of battery in my hive setup is showing as unavailable. Because the binary sensor doesn't have any recollection of any previous state. The issue occurs  when the logic runs async update the flow of that function currently is to get the state first and then check if the device is online, because of this sequence of events it fails.


With the change made now it will check if the device is available first and only if it's available get the state.

Reply:
As you can see here this is one of my sensors when debugging it and it has no state as its offline

```
'hiveID' ='xxxxxxxxx-xxxxxxxx-xxxxxxxxxxx'
'hiveName' ='Back door'
'hiveType' ='contactsensor'
'haType' ='binary_sensor'
'deviceData' =
{'online': False, 'presenceLastChanged': 1693655830425, 'model': 'DWS003', 'version': '05042603', 'manufacturer': 'HiveHome.com', 'upgrade': {'available': False, 'version': '05042603', 'upgrading': False, 'status': 'COMPLETE'}, 'power': 'battery', 'signal': 71, 'battery': 80, 'deviceClass': 'OTHER', 'isSecurityLite': False, 'capabilities': ['INFORMATION', 'RENAME', 'DELETE', 'DEVICE_CLASS', 'SECURITY_DEVICE', 'SECURITY_ZONE', 'SECURITY_TOGGLE', 'SECURITY_TRIGGER']}
'parentDevice' ='xxxxxxxxx-xxxxxxxx-xxxxxxxxxxx'
'isGroup' =False
'device_id' ='xxxxxxxxx-xxxxxxxx-xxxxxxxxxxx'
'device_name' ='Back door'
'haName' ='Back door'
```

This is a snippet from my debugging

Reply:
The point being, shouldn't it be this in that case?

```python
        if self.device["hiveType"] != "Connectivity":
            self._attr_available = (
                self.device["deviceData"].get("online")
                and "state" in self.device["status"]
            )
        else:
            self._attr_available = True

        if self._attr_available:
            self._attr_is_on = self.device["status"]["state"]
```

Reply:
Yep it can be to make it more robust. The current setup is pretty binary if online it has a state and if offline it doesn’t.

 Ill push a commit with the suggested change as it makes sense for it to be more robust.

Reply:
@frenck I have made the change suggested. I decided to look for `status` in device instead of `state`. As the status key is missing as per the error in the description of this PR.